---
title: "Introduction to QuanT"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction to QuanT}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: references.bib
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  warning = FALSE,
  comment = "#>",
  out.width = "80%",
  out.height = "80%",
  dpi = 300
)
```

## Unmeasured heterogeneity in microbiome data and QuanT
Unmeasured technical and biomedical heterogeneity in microbiome data can arise from differential processing and design. Uncorrected for, this can lead to spurious results.  

Currently, no methods are designed to address unmeasured heterogeneity in microbiome studies. Techniques developed for genomic data, such as surrogate variable analysis (SVA) [@leek2007capturing] and remove unwanted variation (RUV) [@risso2014normalization], may be applicable. SVA is a two-step procedure for identifying surrogate variables, a general framework that was adapted by subsequent methods, including ours here. RUV employs factor analysis on control genes (via spike-in experiment or housekeeping gene estimation) to identify hidden heterogeneities. 

However, SVA and RUV may not be ideally suited for microbiome data. First, they do not address the inflated zeros and over-dispersion typical in microbiome data. Second, they detect the underlying structure in the mean expression (log-transformed if needed) while ignoring other critical aspects of the complex data. Third, RUV requires users to specify the number of hidden directions: incorrect specification of these parameters may result in false positives or negatives. 

We propose the Quantile Thresholding (QuanT) approach, a comprehensive non-parametric hidden variable inference method that accommodates the complex distributions of microbial read counts and relative abundances.

## The CRC dataset
We will use the CRC dataset as an example to show how to use the `QuanT` R package. We will also run SVA and RUV and compare them with QuanT based on the CRC dataset.

The CRC dataset originates from the curatedMetagenomicData package [@pasolli2017accessible], which offers standardized, curated human microbiome data across various studies. We analyzed two studies [@feng2015gut; @thomas2019metagenomic] focused on colorectal cancer (CRC), considering distinct patient phenotypes (adenocarcinoma, adenoma, advanced adenoma, carcinoma, and control) as the primary variable of interest, and age and gender as covariates for adjustment. The processed data encompassed 166 genera across 234 samples. 

This dataset is included in the package and is lazy-loaded when loading the package. Let's have a quick look at the data.
```{r}
library(QuanT)

str(crc)
```

As you can see, the list `crc` includes the following components.

1. otu: the absolute abundance(counts) matrix
2. rela: the relative abundance matrix
3. groupid: the group(study) indicator
4. condition: a factor with three levels: adenoma, CRC, control
5. age: a factor with two levels: adult, senior
6. subtype: a factor with five levels: adenoma, carcinoma, adenocarcinoma, advancedadenoma, control
7. gender: a factor with two levels: male, female

## Running QuanT
Now let's run QuanT on this dataset by calling the `QuanT` function. `QuanT` has two arguments: The first is the relative abundance matrix. The second is a matrix containing all known variables, which are phenotype, age, and gender here. For now, we act as if the true group IDs are unknown, and check whether QuanT is able to identify this unmeasured heterogeneity.
```{r}
covariates = model.matrix(~ subtype + age + gender, data=crc)[,-1]
qsv = QuanT(crc$rela, covariates, 
            parallel = TRUE, num.cores = 8)
```

Now we have constructed the QSVs and stored them into the `qsv` object. We can plot the first two QSVs and color the points based on the true group ID.
```{r}
plot(qsv[,1], qsv[,2], col=crc$groupid, 
     cex.lab=0.5, cex.axis=0.5, cex=0.5)
```

The two groups are approximately seperated by the first two QSVs! This indicates that QuanT has succesfully identified the unmeasured heterogeneity.

We can set the argument `visualization` to TRUE when calling the `QuanT` function. This will display the singular values of the (binarized and centered) residual matrix that we obtain through quantile regression.
```{r}
qsv = QuanT(crc$rela, covariates, 
            parallel = TRUE, num.cores = 8, 
            visualization = TRUE)
```

If you have a `phyloseq` object, you could just input it as the first argument of the `QuanT` function. The function will automatically handle it.

## Running SVA and RUV
### Installation
We then run SVA and RUV on the CRC dataset and compare them with QuanT. First install SVA, RUV, and their dependencies if they haven't been installed. We will also need the `ggplot2` and `corrplot` packages for visualization.
```{r message=FALSE}
if (!require("mgcv", quietly = TRUE)) {
 install.packages("mgcv")
}
if (!require("BiocManager", quietly = TRUE)) {
  install.packages("BiocManager")
}
if (!require("sva", quietly = TRUE)) {
  BiocManager::install("sva")
}
if (!require("RUVSeq", quietly = TRUE)) {
  BiocManager::install("RUVSeq")
}
if (!require("ggplot2", quietly = TRUE)) {
 install.packages("ggplot2")
}
if (!require("corrplot", quietly = TRUE)) {
  install.packages("corrplot")
}

library(mgcv)
library(sva)
library(RUVSeq)
library(ggplot2)
library(corrplot)
```

### Running SVA
After installation, we can run SVA on the CRC dataset. `sv` are the surrogate variables(SVs) constructed by SVA.
```{r}
mod = model.matrix(~ subtype + age + gender, data=crc)
mod0 = mod[,1]

svafit = svaseq(t(crc$otu),mod,mod0)
sv = svafit$sv
```

There are eight SVs in total. We can plot the first two SVs and color the points based on the true group ID, just like what we have done for QuanT.
```{r}
plot(sv[,1], sv[,2], col=crc$groupid, 
     cex.lab=0.5, cex.axis=0.5, cex=0.5)
```

The two groups are not well separated by the first two SVs.

## Running RUV
We then run RUV on the CRC dataset. `ruv` are the factors constructed by RUV.
```{r}
design <- model.matrix(~ subtype + age + gender, data=crc)
y <- DGEList(counts=t(crc$otu))
y <- calcNormFactors(y, method="TMM")
y <- estimateGLMCommonDisp(y, design)
y <- estimateGLMTagwiseDisp(y, design)
fit <- glmFit(y, design)
lrt <- glmLRT(fit, coef=2)
empirical = which(p.adjust(lrt$table$PValue,"BH") > 0.05)
ruv <- RUVg(t(crc$otu), empirical, k=11)$W
```

Similar to what we have done for QuanT and SVA, we can plot the first two factors and color the points based on the true group ID.
```{r}
plot(ruv[,1], ruv[,2], col=crc$groupid, 
     cex.lab=0.5, cex.axis=0.5, cex=0.5)
```

RUV also does a good job in separating the two groups based on the first two factors. However, the number of factors needs to be chosen arbitrarily, and an inappropriate number may impair the performance. The original paper chooses 6 factors for a dataset with 128 samples. Based on this choice, we use 11 factors for the CRC data which includes 234 samples.

## Differential abundance analysis
We now conduct differential abundance analysis based on linear regression. The goal of this analysis is to identify the taxa associated with the phenotype while adjusting for the covariates and the hidden variables constructed by each method. For benchmarking, we also performed linear regression adjusted for the true group indicators (as the gold standard) and without any adjustment (as a baseline result). 
```{r}
get_pv = function(rela, surrogates) {
  if (is.null(surrogates)) {
    pv = apply(rela, 2,
               function(x) anova(lm(x~covariates),
                                 lm(x~covariates[,-(1:4)]))$'Pr(>F)'[2])
  } else {
    pv = apply(rela, 2,
               function(x)
                 anova(lm(x~covariates+surrogates),
                       lm(x~covariates[,-(1:4)]+surrogates))$'Pr(>F)'[2])
  }
  return(pv)
}

pv_no_correction = get_pv(crc$rela, NULL)
pv_group = get_pv(crc$rela, crc$groupid)
pv_quant = get_pv(crc$rela, qsv)
pv_sva = get_pv(crc$rela, sv)
pv_ruv = get_pv(crc$rela, ruv)
```

Now we have obtained the p-values from each method. We apply Benjamini-Hochberg procedure to the p-values and use a threshold of 0.05 to determine the significant taxa. Below are the indices of the significant taxa from each method.
```{r}
which(p.adjust(pv_no_correction,"BH") <= 0.05)
which(p.adjust(pv_group,"BH") <= 0.05)
which(p.adjust(pv_quant,"BH") <= 0.05)
which(p.adjust(pv_sva,"BH") <= 0.05)
which(p.adjust(pv_ruv,"BH") <= 0.05)
```
As can be seen from the list above, using QuanT leads to 4 discoveries which are all true discoveries according to the gold standard. SVA leads to many more false discoveries. And RUV leads to only 1 true discovery. We can visualize the results by running the following code to generate a heatmap showing the number of overlapping discoveries between each pair of methods.

```{r}
get_overlap = function(pv, method_name, alpha) {
  n_method = ncol(pv)
  discovery = apply(pv, 2, function(x) which(p.adjust(x,"BH")<=alpha))
  concordance = matrix(NA, nrow=n_method,ncol=n_method)
  for (i in 1:n_method) {
    for (j in i:n_method) {
      if (length(discovery) == 0) {
        concordance[i,j] = 0
      } else if ("list" %in% class(discovery)) {
        concordance[i,j] = length(intersect(discovery[[i]], discovery[[j]]))
      } else {
        concordance[i,j] = length(intersect(discovery[,i], discovery[,j]))
      }
      concordance[j,i] = concordance[i,j]
    }
  }
  rownames(concordance) = colnames(concordance) = method_name
  concordance
}
plot_overlap = function(pv, method_name, alpha=0.05) {
  concordance = get_overlap(pv, method_name, alpha)
  maxconcord = max(concordance)
  corrplot(concordance, 
           method = "color", 
           type = "upper",
           col.lim = c(0,maxconcord),
           # col = rev(COL2('RdYlBu',200)),
           col = COL1("YlGn",200),
           addCoef.col = "black",
           is.corr = FALSE,
           tl.cex = 0.4, number.cex = 0.4,
           tl.srt = 0,
           cl.pos = "n")
}
method_name = c("No Correction",
                "True Group",
                "SVA",
                "RUV",
                "QuanT")
plot_overlap(cbind(pv_no_correction, 
                   pv_group, 
                   pv_sva, 
                   pv_ruv, 
                   pv_quant),
             method_name)
```

## References
