---
title: "Introduction to QuanT"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Introduction to QuanT}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  warning = FALSE,
  comment = "#>",
  out.width = "80%",
  out.height = "80%",
  dpi = 300
)
```

## QuanT
Unmeasured technical and biomedical heterogeneity in microbiome data can arise from differential processing and design. Uncorrected for, this can lead to spurious results.  We propose the Quantile Thresholding (QuanT) approach, a comprehensive non-parametric hidden variable inference method that accommodates the complex distributions of microbial read counts and relative abundances.

## The CRC dataset
We will use the CRC dataset as an example to show how to use the `QuanT` R package. We will also run SVA and RUV and compare them with QuanT based on the CRC dataset.

The CRC dataset originates from the curatedMetagenomicData package, which offers standardized, curated human microbiome data across various studies. We analyzed two studies focused on colorectal cancer (CRC), considering distinct patient phenotypes (adenocarcinoma, adenoma, advanced adenoma, carcinoma, and control) as the primary variable of interest, and age and gender as covariates for adjustment. The processed data encompassed 166 genera across 234 samples. 

This dataset is included in the package and is lazy-loaded when loading the package. Let's have a quick look at the data.
```{r}
library(QuanT)

str(crc)
```

As you can see, the list `crc` includes the following components.

1. otu: the absolute abundance(counts) matrix
2. rela: the relative abundance matrix
3. batchid: the batch(study) indicator
4. condition: a factor with three levels: adenoma, CRC, control
5. age: a factor with two levels: adult, senior
6. subtype: a factor with five levels: adenoma, carcinoma, adenocarcinoma, advancedadenoma, control
7. gender: a factor with two levels: male, female

## Running QuanT
Now let's run QuanT on this dataset by calling the `QuanT` function. `QuanT` has two arguments: The first is the relative abundance matrix. The second is a matrix containing all known variables, which are phenotype, age, and gender here. For now, we act as if the true group IDs are unknown, and check whether QuanT is able to identify this unmeasured heterogeneity.
```{r}
covariates = model.matrix(~ subtype + age + gender, data=crc)[,-1]
qsv = QuanT(crc$rela, covariates)
```

Now we have constructed the QSVs and stored them into the `qsv` object. We can plot the first two QSVs and color the points based on the true group ID.
```{r}
plot(qsv[,1], qsv[,2], col=crc$batchid, 
     cex.lab=0.5, cex.axis=0.5, cex=0.5)
```

The two batches are approximately seperated by the first two QSVs! This indicates that QuanT has succesfully identified the unmeasured heterogeneity.

## Running SVA and RUV
### Installation
We then run SVA and RUV on the CRC dataset and compare them with QuanT. First install SVA, RUV, and their dependencies if them haven't been installed.
```{r message=FALSE}
if (!require("mgcv", quietly = TRUE)) {
 install.packages("mgcv")
}
if (!require("BiocManager", quietly = TRUE)) {
  install.packages("BiocManager")
}
if (!require("sva", quietly = TRUE)) {
  BiocManager::install("sva")
}
if (!require("RUVSeq", quietly = TRUE)) {
  BiocManager::install("RUVSeq")
}

library(mgcv)
library(sva)
library(RUVSeq)
```

### Running SVA
After installation, we can run SVA on the CRC dataset. `sv` are the surrogate variables(SVs) constructed by SVA.
```{r}
mod = model.matrix(~ subtype + age + gender, data=crc)
mod0 = mod[,1]

svafit = svaseq(t(crc$otu),mod,mod0)
sv = svafit$sv
```

## Running RUV
We then run RUV on the CRC dataset. `ruv` are the factors constructed by RUV.
```{r}
design <- model.matrix(~ subtype + age + gender, data=crc)
y <- DGEList(counts=t(crc$otu))
y <- calcNormFactors(y, method="TMM")
y <- estimateGLMCommonDisp(y, design)
y <- estimateGLMTagwiseDisp(y, design)
fit <- glmFit(y, design)
lrt <- glmLRT(fit, coef=2)
empirical = which(p.adjust(lrt$table$PValue,"BH") > 0.05)
ruv <- RUVg(t(crc$otu), empirical, k=11)$W
```

## Differential abundance analysis
We now conduct differential abundance analysis based on linear regression. The goal of this analysis is to identify the taxa associated with the phenotype while adjusting for the covariates and the hidden variables constructed by each method. For benchmarking, we also performed linear regression adjusted for the true group indicators (as the gold standard) and without any adjustment (as a baseline result). 
```{r}
get_pv = function(rela, surrogates) {
  if (is.null(surrogates)) {
    pv = apply(rela, 2,
               function(x) anova(lm(x~covariates),
                                 lm(x~covariates[,-(1:4)]))$'Pr(>F)'[2])
  } else {
    pv = apply(rela, 2,
               function(x)
                 anova(lm(x~covariates+surrogates),
                       lm(x~covariates[,-(1:4)]+surrogates))$'Pr(>F)'[2])
  }
  return(pv)
}

pv_no_correction = get_pv(crc$rela, NULL)
pv_batch = get_pv(crc$rela, crc$batchid)
pv_quant = get_pv(crc$rela, qsv)
pv_sva = get_pv(crc$rela, sv)
pv_ruv = get_pv(crc$rela, ruv)
```

Now we have obtained the p-values from each method. We apply Benjamini-Hochberg procedure to the p-values and use a threshold of 0.05 to determine the significant taxa. Below are the indices of the significant taxa from each method.
```{r}
which(p.adjust(pv_no_correction,"BH") <= 0.05)
which(p.adjust(pv_batch,"BH") <= 0.05)
which(p.adjust(pv_quant,"BH") <= 0.05)
which(p.adjust(pv_sva,"BH") <= 0.05)
which(p.adjust(pv_ruv,"BH") <= 0.05)
```
As can be seen from the list above, using QuanT leads to 4 discoveries which are all true discoveries according to the gold standard. SVA leads to many more false discoveries. And RUV leads to only 1 true discovery..

## Proportion of overlapping discoveries
Other than thresholding Benjamini-Hochberg adjusted p-values at 0.05, we can also sort the p-values and consider the top K significant taxa with a varying K. We can calculate the proportions of overlapping taxa in the top K detected differentially abundant taxa between each analysis and the gold standard analysis in the CRC dataset. Here, the method adjusting for true group or study indicator is considered as gold standard.
```{r}
library(ggplot2)
truerank = rank(pv_batch, ties.method = "min")
get_overlap = function(pv) {
  overlap = rep(NA,20)
  for (k in 1:20) {
    dis = which(rank(pv, ties.method = "min") <= k)
    truetopK = which(truerank <= k)
    overlap[k] = mean(dis%in%truetopK)
  }
  return(overlap)
}

method_name = c("No Correction",
                "True Group",
                "SVA",
                "RUV",
                "QuanT")
df = data.frame(overlap = c(get_overlap(pv_no_correction), 
                             get_overlap(pv_batch), 
                             get_overlap(pv_sva), 
                             get_overlap(pv_ruv), 
                             get_overlap(pv_quant)),
                 method = rep(method_name,each=20),
                 K = rep(1:20,length(method_name)))
df$method = factor(df$method, method_name)
pic_overlap = ggplot(df, aes(x = K,
             y = overlap,
             color = method)) +
  geom_line(lwd = 0.6) +
  theme_bw() +
  scale_color_manual(values = c("#7FC97F","#BEAED4","#FDC086","#386CB0","#F0027F")) +
  labs(color = "Method",
       shape = "Method",
       linetype = "Method",
       x = "Top K Detected Taxa",
       y = "Proportion of overlapping Discoveries") +
    theme(legend.position="bottom",
          text = element_text(size = 6)) +
  guides(color = guide_legend(nrow = 2))
plot(pic_overlap)
```

As can be seen from the figure, QuanT is most consistent with the gold standard in terms of top K differential abundant taxa.
